test/integration/endpointslice/endpointslice_terminating_test.go

```diff
@@ -0,0 +1,337 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslice
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/controller/endpointslice"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/test/integration/framework"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// TestEndpointSliceTerminating tests that terminating pods are NOT included in EndpointSlice when
+// the feature gate EndpointSliceTerminatingCondition is off. If the gate is on, it tests that
+// terminating endpoints are included but with the correct conditions set for ready, serving and terminating.
+func TestEndpointSliceTerminating(t *testing.T) {
+	testcases := []struct {
+		name            string
+		podStatus       corev1.PodStatus
+		endpoints       []discovery.Endpoint
```

Comment:
Tiny nit: It's easier for me to understand test cases when there's some differentiation between input and output. It might be helpful to use `expectedEndpoints` as a name here.